AuthenticatedClient overhaul, docstrings, and testing

acontry opened the pull request:
This is an overhaul of the AuthenticatedClient. A summary:
* Consistent interface for all endpoints (#60)
* Paginated requests have been abstracted away in the form of generators (#64)
* Docstrings for all methods (#56)
* Tests for as many authenticated endpoints as possible. Sandbox API credentials need to be supplied to run the tests (instructions in readme). The sandbox API is a little flaky, so some tests randomly fail. Not the best testing methodology, but it exercises most of the functions at least.

An AuthenticatedClient object now makes all HTTP calls through a single requests.Session. This may give a performance boost when making many calls to the GDAX API and should remain thead-safe. All HTTP requests are made through helper functions for cleaner code, and should also make implementing error handling simple to do in the future. These changes are implemented in PublicClient as well.

This PR has breaking changes in the interface, so we can discuss how to integrate without losing sync between Github and pypi again. Maybe push this to a version branch, and then merge to master when ready to update pypi?

Thanks to @Jigsaw-Jams for writing most of the docstrings!

PaulMest commented:
Looks like a pretty solid PR. I'll try to find some time to test it and make comments if you'd like some feedback.

acontry commented:
@PaulMest That would be good if someone else could try out the new tests! I never tried running them on a fresh GDAX sandbox account, so that would be a good check.

danpaquin commented:
I like all of this, with the exception of refactoring orders.  While I agree we may add some more user friendly "stop_limit_order" methods, we should not remove buy or sell.  I almost feel like we should replace them all with "place_order" that other order types may be built on.   My goal is to keep this wrapper as true to the official GDAX API doc's as possible.  Does this make sense?  Please let me know your thoughts.

acontry commented:
@danpaquin I'm not sure I fully followed you - would adding `buy` and `sell` as wrappers of `place_order` work for you? 

Right now `place_order` is the main function with all the logic, and `place_market_order`, `place_limit_order`, and `place_stop_order` simply call it. My motivation in splitting the order function up was to make the code more user-friendly. Each order type has a large number of parameters, so having separate functions with named parameters and corresponding docstrings should hopefully make it easier to build a proper order without having to consult the GDAX API spec. I definitely agree with trying to stick to the spec - that was my intent!

mrgordon commented:
Can we at least merge the readme update soon? It says you can use `product_id` when initializing `AuthenticatedClient` but it fails on the pip version and the GitHub version.

dmitrytorba commented:
It would be nice to have the same type of pagination in public_client as well. Like for `get_product_trades()`, for example. Maybe ` _send_paginated_message()` can be shared between the two?

dmitrytorba commented on gdax/authenticated_client.py:
```diff
@@ -15,281 +15,998 @@
-        # r.raise_for_status()
+        """  Get your 30-day trailing volume for all products.
+
+        This is a cached value that’s calculated every day at midnight UTC.
```
The apostrophe in `that’s` seems to have weird encoding, and its causing an error: `SyntaxError: Non-ASCII character '\xe2'`

danpaquin commented:
I love the pagination, docstrings and all other improvements, but I would not like to break other backwards compatibility.  We can keep as it, but add back buy() and sell() methods for backwards compatibility.

Thanks!

acontry commented:
@dmitrytorba Thanks for the review and suggestion, I implemented it in latest commit.

@danpaquin Thanks for the comment - I added back buy() and sell(), they should be backwards compatible now. 

Now both PublicClient and AuthenticatedClient use  `_send_message` and `_send_paginated_message` for every API call. Clean and consistent across the whole library.

dmitrytorba commented:
@acontry, awesome work! 
looks like there are conflicts with master

acontry commented:
@dmitrytorba Thanks! Updated this branch one more time...

danpaquin commented:
I am nervous to implement such drastic changes in one PR.  Backwards comparability is my main concern regarding order's.  I would like to keep the wrapper as close to bare-bones as possible so users can look at the wrapper README alongside the GDAX API Reference.  Higher level methods can be added in developers' inherited classes.

acontry commented:
@danpaquin Backwards compatibility is there for orders. I understand this is a big PR now - if these changes are still desired, I would suggest creating a dev branch where these changes can be merged. That makes it easier for them to be fully vetted before merging into master.

Bkyn commented:
Just a lurker passing by: I can see the awesome work done on this module in general and in particular in this PR.  
From a newbie point of view, the documentation added to AuthenticatedClient is great!  
I hope you can merge it soon with the master so that I can develop with it.

danpaquin commented:
Hi Bkyn, thank you for sharing your feedback.  I will not merge to master since many methods have been converted to generators which may break older code.  I love the docstrings and refactors, but this PR has become too big of a change to push all at once. I will work on creating a dev branch to pull this into.

acontry commented:
@danpaquin sounds good to me! I just pointed this PR at development branch for you.
